api/protocol.go

```diff
@@ -52,68 +50,148 @@ type Response struct {
-// Response
-func ReadMessage(reader io.Reader, msg interface{}) error {
-	buf := make([]byte, headerLength)
-	n, err := reader.Read(buf)
+func maxOpcodeForFrameType(t FrameType) int {
+	switch t {
+	default:
+		fallthrough
+	case TypeCommand:
+		return int(CmdMax)
+	case TypeResponse:
+		return int(CmdMax)
+	case TypeStream:
+		return int(StreamMax)
+	case TypeNotification:
+		return int(NotificationMax)
+	}
+}
+
+// ReadFrame reads a full frame (header and payload) from r.
+func ReadFrame(r io.Reader) (*Frame, error) {
+	// Read the header.
+	buf := make([]byte, minHeaderLength)
+	n, err := r.Read(buf)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if n != headerLength {
-		return errors.New("couldn't read the full header")
+	if n != minHeaderLength {
+		return nil, errors.New("frame: couldn't read the full header")
 	}
 
-	header := header{
-		length: binary.BigEndian.Uint32(buf[0:4]),
-		flags:  binary.BigEndian.Uint32(buf[4:8]),
+	// Decode it.
+	frame := &Frame{}
+	header := &frame.Header
+	header.Version = int(binary.BigEndian.Uint16(buf[0:2]))
+	if header.Version < 2 || header.Version > Version {
+		return nil, fmt.Errorf("frame: bad version %d", header.Version)
+	}
+	header.HeaderLength = int(buf[2]) * 4
+	header.Type = FrameType(buf[6] & 0xf)
+	flags := buf[6] & 0xf0
+	if flags&flagInError != 0 {
+		header.InError = true
+	}
+	if header.Type >= TypeMax {
+		return nil, fmt.Errorf("frame: bad type %s", header.Type)
 	}
+	header.Opcode = int(buf[7])
+	if header.Opcode >= maxOpcodeForFrameType(header.Type) {
+		return nil, fmt.Errorf("frame: bad opcode (%d) for type %s", header.Opcode,
+			header.Type)
+	}
+	header.PayloadLength = int(binary.BigEndian.Uint32(buf[8:12]))
 
+	// Read the payload.
 	received := 0
-	need := int(header.length)
-	data := make([]byte, need)
+	need := header.HeaderLength - minHeaderLength + header.PayloadLength
+	payload := make([]byte, need)
 	for received < need {
-		n, err := reader.Read(data[received:need])
+		n, err := r.Read(payload[received:need])
 		if err != nil {
-			return err
+			return nil, err
 		}
 
 		received += n
 	}
 
-	err = json.Unmarshal(data, msg)
-	if err != nil {
-		return err
-	}
+	// Skip the bytes part of a bigger header than expected to just keep
+	// the payload.
+	frame.Payload = payload[header.HeaderLength-minHeaderLength : need]
 
-	return nil
+	return frame, nil
 }
 
-// WriteMessage writes a message into writer. A message is either a Request for
-// a Response
-func WriteMessage(writer io.Writer, msg interface{}) error {
-	data, err := json.Marshal(msg)
-	if err != nil {
-		return err
+const (
+	flagInError = 1 << (4 + iota)
+)
+
+// WriteFrame writes a frame into w.
+//
+// Note that frame.Header.PayloadLength dictates the amount of data of
+// frame.Payload to write, so frame.Header.Payload must be less or equal to
+// len(frame.Payload).
+func WriteFrame(w io.Writer, frame *Frame) error {
```

Comment:
Ok, while ideally I want to avoid the copy in your patch with the proposed api change, I think it's more important to have WriteFrame working from multiple goroutines. I've included your patch in this branch (thanks!) and opened a new issue ( https://github.com/clearcontainers/proxy/issues/17) to track the removal of the copy.